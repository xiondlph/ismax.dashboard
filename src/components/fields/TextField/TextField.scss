$dashboard-textfield-trigger-error-color: #cd5c5c;
$dashboard-textfield-border-color: #adb3b8;
$dashboard-textfield-border-focused-color: #6495ed;
$dashboard-textfield-border-error-color: $dashboard-textfield-trigger-error-color;

.dashboard-field-text {
  border: .5px solid $dashboard-textfield-border-color;
  background-color: white;
  border-radius: 3px;
  position: relative;
  padding-right: 5px;
  display: flex;
  height: 37px;
  margin: 10px;

  &_focused {
    border: .5px solid $dashboard-textfield-border-focused-color;

    .dashboard-field-text-label {
      color: $dashboard-textfield-border-focused-color;
    }
  }

  &_filled {
    .dashboard-field-text-label {
      font-size: 10px;
      top: 3px;
    }
  }

  &_error {
    border: .5px solid $dashboard-textfield-border-error-color;
  }

  &-input {
    background-color: transparent;
    padding: 16px 0 3px 10px;
    outline: none;
    border: none;
    height: 18px;
    flex: 1;
  }

  &-label {
    color: $dashboard-textfield-border-color;
    transition: top .2s, font-size .2s;
    position: absolute;
    font-size: 14px;
    left: 10px;
    top: 17px;
  }

  &-trigger {
    color: $dashboard-text-color-gray;
    text-align: center;
    line-height: 37px;
    font-size: 15px;
    display: block;
    height: 37px;
    width: 20px;

    &_error {
      color: $dashboard-textfield-trigger-error-color;
    }

    &-icon {
      cursor: pointer;
    }
  }
}