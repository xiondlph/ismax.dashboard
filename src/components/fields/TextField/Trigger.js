/**
 * Стандартный триггер
 */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

class Trigger extends Component {
    onIconClick() {
        const { handler } = this.props

        if (typeof handler === 'function') {
            handler()
        }
    }

    render() {
        const { tip, icon, type } = this.props

        return (
            <div className={classNames({
                'dashboard-field-text-trigger': true,
                [`dashboard-field-text-trigger_${type}`]: !!type
            })}>
                <span
                    className={`dashboard-field-text-trigger-icon fa fa-${icon}`}
                    data-tip={tip}
                    data-type={type}
                    onClick={::this.onIconClick}
                />
            </div>
        )
    }
}

Trigger.propTypes = {
    tip: PropTypes.string.isRequired,
    icon: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    handler: PropTypes.func
}

export default Trigger