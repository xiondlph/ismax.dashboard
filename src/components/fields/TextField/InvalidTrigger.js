/**
 * Триггер ошибки валидации
 */

import React from 'react'
import PropTypes from 'prop-types'
import Trigger from './Trigger'

const InvalidTrigger = props => {
    const { error } = props

    return <Trigger tip={error} icon='exclamation-circle' type='error'/>
}

InvalidTrigger.propTypes = {
    error: PropTypes.string.isRequired
}

export default InvalidTrigger