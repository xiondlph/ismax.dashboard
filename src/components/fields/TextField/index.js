/**
 * Компонент текстового поля
 */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import ReactTooltip from 'react-tooltip'
import Trigger from './Trigger'
import InvalidTrigger from './InvalidTrigger'

import './TextField.scss'

export default class TextField extends Component {
    constructor(props) {
        super(props)

        this.state = {
            filled: false
        }
    }

    componentDidUpdate() {

        // Обновим тултипы
        ReactTooltip.rebuild();

        this.updateLabelPosition()
    }

    componentDidMount() {
        this.updateLabelPosition()
    }

    updateLabelPosition() {
        const { input: { value } } = this.props

        if (this.state.filled !== !!value) {
            this.setState({
                filled: !!value
            })
        }
    }

    onChange(e) {
        const { input: { onChange } } = this.props

        onChange(e.target.value)
    }

    triggers() {
        const { triggers } = this.props
        let items = []

        for(let index in triggers) {
            items.push({
                key: index,
                ...triggers[index]
            })
        }

        return items.map(item => (<Trigger
                key={item.key}
                tip={item.tip}
                icon={item.icon}
                type={item.type}
                handler={item.handler}
            />)
        )
    }


    render() {
        const
            {
                input: { value, onFocus, onBlur },
                meta: { error, active },
                label
            } = this.props

        return (
            <div className={classNames({
                'dashboard-field-text': true,
                'dashboard-field-text_focused': active,
                'dashboard-field-text_filled': this.state.filled,
                'dashboard-field-text_error': error
            })}>
                <input
                    className='dashboard-field-text-input'
                    type='text'
                    value={value}
                    onChange={ ::this.onChange }
                    onFocus={ onFocus }
                    onBlur={ onBlur }
                />
                <div className='dashboard-field-text-label'>{label}</div>
                {error && (<InvalidTrigger error={error}/>)}
                {::this.triggers()}
            </div>
        )
    }
}

TextField.propTypes = {
    type: PropTypes.string.isRequired,
    label: PropTypes.string,
    input: PropTypes.object.isRequired,
    meta: PropTypes.object.isRequired,
    trigger: PropTypes.object
}