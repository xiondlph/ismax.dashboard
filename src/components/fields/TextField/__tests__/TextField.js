import React from 'react'
import renderer from 'react-test-renderer'
import { configure, shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import TextField from '../index'

configure({ adapter: new Adapter() })

describe('TextField', () => {
    test('Проверка TextField снапшота', () => {
        const
            type = 'text',
            label = 'Email',
            input = {
                value: 'user@email.com',
                onChange: jest.fn(),
                onFocus: jest.fn(),
                onBlur: jest.fn()
            },
            meta = {},
            triggers = {
                email: {
                    info: {
                        tip: 'info',
                        icon: 'question'
                    }
                }
            },
            component = renderer.create(
                <TextField type={type} label={label} input={input} meta={meta} trigger={triggers.email} />
            )

        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })

    test('Проверка невалидного значения в TextField', () => {
        const
            type = 'text',
            label = 'Email',
            input = {
                value: 'user@email.com',
                onChange: jest.fn(),
                onFocus: jest.fn(),
                onBlur: jest.fn()
            },
            meta = {},
            triggers = {
                email: {
                    info: {
                        tip: 'info',
                        icon: 'question',
                        type: 'info'
                    }
                }
            },
            textField = mount(
                <TextField type={type} label={label} input={input} meta={meta} triggers={triggers.email} />
            );

        expect(textField.find('.dashboard-field-text_error').exists()).toEqual(false)

        textField.setProps({
            meta: {
                error: 'invalid'
            }
        })

        expect(textField.find('.dashboard-field-text_error').exists()).toEqual(true)

        // Проверка наличия триггиров
        expect(textField.find('.dashboard-field-text-trigger_error').length).toBe(1)
        expect(textField.find('.dashboard-field-text-trigger_info').length).toBe(1)
    })

    test('Проверка поведения на наличия значения в TextField', () => {
        const
            type = 'text',
            label = 'Email',
            input = {
                value: '', // Пустое значение
                onChange: jest.fn(),
                onFocus: jest.fn(),
                onBlur: jest.fn()
            },
            meta = {},
            triggers = {
                email: {
                    info: {
                        tip: 'info',
                        icon: 'question'
                    }
                }
            },
            textField = shallow(
                <TextField type={type} label={label} input={input} meta={meta} trigger={triggers.email} />
            );

        expect(textField.find('.dashboard-field-text_filled').exists()).toEqual(false)

        textField.setProps({
            input: {
                value: 'user@email.com'
            }
        })

        expect(textField.find('.dashboard-field-text_filled').exists()).toEqual(true)
    })
})