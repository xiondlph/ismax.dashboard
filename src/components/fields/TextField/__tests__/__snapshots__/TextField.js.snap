// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Проверка TextField снапшота 1`] = `
<div
  className="dashboard-field-text dashboard-field-text_filled"
>
  <input
    className="dashboard-field-text-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="user@email.com"
  />
  <div
    className="dashboard-field-text-label"
  >
    Email
  </div>
</div>
`;
